### DISCUSSION TITLE
-

### DISCUSSION THREAD

**Discussion by IcyFold** [gucci-on-fleek/lockdown-browser#20] [Q&A]

**Comment by gucci-on-fleek** [comment]
> Do you know if Respondus has the ability to mitm your web traffic

I doubt it. To do that they'd need to install a root CA on your computer. But Windows doesn't let programs do that without asking for your permission, and I've never seen a permission popup before.

But to be safe, you can check your list of installed certificates by running

```powershell
certmgr.msc
```

Look for anything with either no name, an obviously randomly-generated name, or something with "Respondus" or "Lockdown Browser" in the name.

Another solution would be to use Firefox; it uses its own internal certificate store, so installing a Windows certificate won't affect it.

(and nearly everything uses HTTPS these days, so a MITM without installing a root certificate would be pointless)

> or whether it sets up a local proxy or has any packet capture functions?

Also unlikely. You can try [running something like this to check](https://superuser.com/a/1372318):

```powershell
netsh wfp show filters
```

There are also a few other ways that it could have installed a network filter, but I don't know how to look for those.

> Specifically, I had a few Chrome tabs running in the background with apps like Whatsapp web open and have received a few messages.

I would be surprised if the _Browser_ was intercepting the traffic from unrelated apps, but the _Browser_ is definitely capable of taking screenshots and uploading them to Respondus + your school. If Chrome was just minimized, it could easily take a screenshot. If it were closed (with no open windows) or if you were on another tab, then it's much less likely that they would have access to anything.

But regardless, if you're running the _Browser_ inside the Sandbox, it certainly can't access anything on your host computer.

**Comment by gucci-on-fleek** [comment]
> Do you know how to check the date on which a Root CA was installed? I can only see the valid dates, but not the installed dates.

Using anything other than a locally-generated certificate would be wildly irresponsible (see [Superfish](https://en.wikipedia.org/wiki/Superfish#Lenovo_security_incident)), so the install date would likely be the same as when you installed the _Browser_. It could also be something suspicious, like "Jan 1 2000", "Jan 1 1970", etc. 

> Also, do you know if it is able to check my search/internet history?

Could it—yes, quite easily. Does it—very unlikely. Anything privacy-invading that the _Browser_ does needs to be either defensible as necessary to secure the exam or as a result of sloppy coding practices. I can't see how they could defend accessing your search history as either.

> Out of curiousity--I had assumed that the Browser was running inside a sandbox of its own already, why is there a need to impose another sandbox around the browser?

Because that's the entire point of this project?

![Screenshot from 2023-06-30 02-49-21](https://github.com/gucci-on-fleek/lockdown-browser/assets/49086429/9b01db19-4524-4330-8cd9-8d8029ef9708)

If you're talking about a sandbox beyond the one implemented here; then no, that's not necessary.

**Reply by gucci-on-fleek** [reply]
I wouldn't be surprised if Respondus Monitor monitored network requests while it was active, but since pretty much everything is encrypted these days, it would only be able to know which servers you are connecting to. 

If it were monitoring WhatsApp for example, it could know what time you send/recieved a message, but it wouldn't know the message contents or who you're communicating with. In this example, this is mostly harmless, although still not great.

But if you opened up a website like `www.how-to-cheat-on-exams.com`, then it could know that you opened up this specific site. It wouldn't know the exact URL or what you did on the site though. Here, just knowing the domain alone is potentially incriminating.

And the _Browser_ could theoretically do this all the time once it's installed, but I would be very surprised if it ever did this outside of an active exam. Respondus is unethical, and arguably corrupt/evil, but they're not criminals. 

Regardless, all of this is moot if you're using the Sandbox tool from this repo.

**Reply by gucci-on-fleek** [reply]
> (the app's ability to take screenshots)--do you know when the browser takes screenshots (e.g. periodically, or upon a certain event), and do you know where the screenshots are stored?

For Respondus Monitor, I'd assume that it takes screenshots continuously. For just the plain _Browser_, it would either be random, triggered by "suspicious activity", or it might never happen.

Any screenshots would be stored on Respondus's servers.

> Also, on a side note, I reaccessed the browser to test it out and closed it halfway. It seems that sometimes when I close it, the browser doesn't fully close--there is no icon on my taskbar/system tray, but somehow it is still running in task manager. I can't replicate this with consistency though. Do you know what is happening here,

This is definitely not the intended behaviour, which makes it a bug.

> do you know if the browser is in such a state it will still continue to record information/take screenshots?

Well once you've hit a bug, you can't really predict anything accurately about a piece of software. So who knows?

**Reply by gucci-on-fleek** [reply]
> Do you know if the screenshotting function is nested within the CEF browser itself or is it within the .exe outside of the CEF browser?

CEF is "Chromium Embedded Framework", i.e. Google Chrome. All of the "lockdown" functionality is implemented in `LockdownBrowser.exe` or the accompanying `.dll`.

> Also, how would Respondus monitor the network requests?

A network filter, an Miniport Adapter, a DNS server/hook, a global proxy server, by injecting a `.dll` into every running process, a kernel driver, by hooking the network adapter, by inspecting the list of open sockets for each running process, etc. Too many possibilities to list.

Of these, I'd say only the filter, miniport, inspecting the list of sockets, or proxy server are likely, and I/someone would have noticed before if it was installing a global proxy.

**Reply by gucci-on-fleek** [reply]
> From what I read, the Lockdown browser is a CEF browser and is meant to be embedded in an application (the lockdown browser.exe in this case). Could it mean that when I perform a three finger swipe, or when I close the browser manually, the browser instance ceases to run but the application still continues to run in the background?

"embedded" here means that the CEF `.dll`s run in the same address space as the embedding process, i.e. they run in the same process. So it wouldn't really be possible to separate the _Browser_ processes from the CEF processes like that.

However, I believe that the _Browser_ does run multiple processes. It would certainly be possible for all the visible windows to be closed but there are still some background processes running. This is generally pretty easy to test though—the same process that implements the spying also implements blocking the other processes. So if you are able to open Task Manager without the _Browser_ immediately closing it, then the _Browser_ is probably (but not certainly) completely closed.

I'm pretty sure that the _Browser_ doesn't set anything to run at startup though, so the easiest way to be sure that everything is fully closed-out is to just reboot. (Or, you know, run it in a sandbox of some sort :) )

**Reply by gucci-on-fleek** [reply]
Probably, although I haven't tested any of this recently so I can't be certain.

**Reply by gucci-on-fleek** [reply]
Well it's been almost 3 years since I initially researched all of this and made the tool in this repo,  2 years since I graduated from the school that required the _Browser_, and over a year since I've had access to a computer running Windows. So quite awhile.

But I've had users confirm that this tool still works (see #21 for the most recent example), so the _Browser_ couldn't have changed too much without breaking all of this.

**Reply by gucci-on-fleek** [reply]
> Is it possible that Lockdown Browser is hijacking Avast's hooking functions for itself to hook the user's device for information?

The _Browser_ isn't loading Avast's `.dll`s, Avast is loading it's `.dll`s into the _Browser_. If you open Notepad, you'll probably see the same thing. 

This kind of stuff isn't terribly well-documented, but see the following for a bit more information:

- https://hacks.mozilla.org/2023/03/letting-users-block-injected-third-party-dlls-in-firefox/
- https://firefox-source-docs.mozilla.org/widget/windows/blocklist.html
- https://blog.chromium.org/2017/11/reducing-chrome-crashes-caused-by-third.html

**Reply by gucci-on-fleek** [reply]
> I opened procmon with other apps like Chrome and they do not seem to exhibit similar behaviour.

Chrome specifically blocks antivirus from doing stuff to it. See the third link in the above post. You probably don't see any firewall stuff with Notepad because it can't access the internet, so there's no need for Avast to do firewall stuff to it.

> Also, I noted that Lockdown Browser loads wininet.dll, dnsapi.dll, IPHLPAPI.dll and ws2_32.dll. From my understanding, they allow applications to query the extant established TCP connections.

Sure, they let the program do that. But they're also required to access the internet at all, which is kind of important for a Browser :)

For example, `wininet.dll` exports [`HttpOpenRequest`](https://learn.microsoft.com/en-us/windows/win32/api/wininet/nf-wininet-httpopenrequesta). All this function does is make an HTTP request, which is pretty innocuous for a browser (but would be really suspicious for Notepad or similar)

> I went on Hybrid Analysis to see the functions imported and the relevant functions did not show up but I thought that the functions could be called dynamically and would not show up as an imported function in Ghidra etc.

[The _Browser_ does some funky dynamic import stuff for `EnumProcesses`](https://github.com/gucci-on-fleek/lockdown-browser/issues/4#issuecomment-1159352265), but I'd expect all the other functions to be in the import table. Maybe they've done some weird anti-debugging stuff to hide it?

Also, you'll probably find it easier to inspect the _Browser_ code by using [Process Monitor](https://learn.microsoft.com/en-us/sysinternals/downloads/procmon) and [`traceapi`](https://github.com/microsoft/Detours/tree/main/samples/traceapi) instead of using Ghidra.

**Reply by gucci-on-fleek** [reply]
> If I use traceapi and Browser loads an image of the .dll and calls functions from the loaded image, would traceapi still be able to track these function calls?

Probably – `traceapi` should show any call of a Windows API function. Kinda like a higher level Procmon (which only shows syscalls).

> Also, I understand that Process Monitor only allows me to view how Browser interacts with various files and the registry. [...] and how can I see if Browser is trying to obtain information on extant and established TCP connections in procmon?

It looks like you'd usually use a function like [`GetExtendedTcpTable`](https://learn.microsoft.com/en-us/windows/win32/api/iphlpapi/nf-iphlpapi-getextendedtcptable) to get a list of open connections. It looks like this function ultimately reads from the file `\Device\Tcp`

https://github.com/reactos/reactos/blob/ed4e3f56111755dc705bb05e76b0e0a95ee38d53/dll/win32/iphlpapi/iphlpapi_main.c#L946

https://github.com/reactos/reactos/blob/ed4e3f56111755dc705bb05e76b0e0a95ee38d53/sdk/lib/tdilib/handle.c#L10

I'd expect that Procmon would either show this read as a special "get_network" call, or as a regular file reading operation.

> How can I view the Browser code behind that interaction

You can't with this method. But what the _Browser_ does itself here isn't terribly interesting, all that we care about is how it interacts with the rest of the system. And Procmon+`traceapi` can show all that.

**Comment by IcyFold** [comment]


**Reply by gucci-on-fleek** [reply]
> In the meantime, on the Browser's ability to screenshot--how did you figure out that it could take screenshots?

It _can_ take screenshots in the same sense that any Windows application is capable of taking screenshots. And I thought that it _did_ take screenshots sometimes, but https://archive.today/WX7lw suggests that it doesn't.

**Reply by gucci-on-fleek** [reply]
> I didn't know that Window applications had that ability! Is this some sort of in-built function into all Windows application?

It's built in to Windows, and any application can use that API, so it's kind of built in to all applications.

> Interesting, were there any characteristics or specific performance traits/incidents of Browser which led you to suggest that it did take screenshots?

I know of other exam monitoring software that takes screenshots, and since _Respondus Monitor_ (which is built in to the _Browser_) records a video from your webcam and the audio from your microphone, taking screenshots too wouldn't be surprising.

And the Browser _might_ still take screenshots, although it seems fairly unlikely based off of reading their website.

**Reply by gucci-on-fleek** [reply]
> I found that Browser.exe seemingly uses GDI32 (particularly bitblt) to render webpages (e.g. when I click on a webpage in Browser, functions from GDI32 get used). To my mind this seems unusual since there are other better options to render webpages, such as the built-in Blink engine in the libcef.dll folder.

You always (as far as I'm aware) need GDI to put anything on a screen in Windows. The _Browser_ itself displays a few dialog boxes, so they need GDI for that. And `libcef.dll` will ultimately use GDI to display stuff too, so you're maybe seeing the GDI calls that it's making.

> How do I know if Browser is really using GDI32 to render a webpage instead of using the opportunity to take screenshots of my screen?

You could run the _Browser_ under a debugger or with the API tracer while you write you exam and see if it tries to take a screenshot. But this is somewhat risky—the debugger or tracer might do something that alerts the _Browser_ to its presence, which will cause the _Browser_ to exit and send a message to your teacher. And unless you're in a Computer Science program, good luck explaining that.

The other alternative is to use the sandbox in this repo. The _Browser_ could still take a screenshot, but it will only be able to access the inside of the sandbox where there's nothing to be seen.

I guess another option would be to completely disable taking screenshots. I'm sure that there's a group policy setting that lets you do this. And another option would be to somehow manipulate the _Browser_'s process security token to deny it from accessing other windows.
